File: packages/contracts/python/butterflylens/contracts/geographic_impact.py
"""Australian geographic-impact wire declarations."""

from __future__ import annotations

from typing import Literal, TypedDict


GEOGRAPHIC_IMPACT_CELL_SCHEMA_VERSION = (
    "butterflylens-geographic-impact-cell:v1.0.0"
)
GEOGRAPHIC_IMPACT_SNAPSHOT_SCHEMA_VERSION = (
    "butterflylens-geographic-impact-snapshot:v1.0.0"
)
GEOGRAPHIC_IMPACT_QUERY_SCHEMA_VERSION = (
    "butterflylens-geographic-impact-query:v1.0.0"
)
EVIDENCE_COUNT_STATUSES = (
    "available",
    "unavailable",
    "withheld",
    "not_applicable",
)


class EvidenceCount(TypedDict):
    status: str
    value: int | None
    reason: str | None


class ImpactFlag(TypedDict):
    status: Literal["available", "unavailable"]
    value: bool | None
    reason: str | None


class GeographicImpactCell(TypedDict):
    schema_version: Literal["butterflylens-geographic-impact-cell:v1.0.0"]
    cell_id: str
    grid: Literal["H3"]
    h3_version: str
    h3_resolution: int
    project_id: str
    run_id: str
    snapshot_mode: Literal["live", "submitted"]
    accepted_taxon_key: str
    ala_snapshot_fingerprint: str | None
    flickr_snapshot_fingerprint: str | None
    provider_union_fingerprint: str | None
    review_projection_fingerprint: str | None
    quality_snapshot_fingerprint: str | None
    counts: dict[str, EvidenceCount]
    impact: dict[str, ImpactFlag]
    nearest_ala_evidence_distance: dict[str, object]
    latest_ala_event_date: str | None
    latest_flickr_event_date: str | None
    data_deficiency_state: str
    public_geometry: dict[str, object]
    evidence_fingerprints: list[str]
    cell_fingerprint: str
    scientific_claim_allowed: Literal[False]


class GeographicImpactSnapshot(TypedDict):
    schema_version: Literal["butterflylens-geographic-impact-snapshot:v1.0.0"]
    snapshot_id: str
    project_id: str
    run_id: str
    mode: Literal["live", "submitted"]
    country_code: Literal["AU"]
    status: Literal["available", "stale", "unavailable"]
    generated_at: str
    last_updated_at: str
    submitted_source_commit: str | None
    worker_heartbeat_fingerprint: str | None
    cell_schema_version: Literal["butterflylens-geographic-impact-cell:v1.0.0"]
    cell_count: int
    cell_artifact_checksum: str
    cell_artifact_fingerprint: str
    query_fingerprint: str
    map_projection_fingerprint: str
    blockers: list[str]
    append_only_revision: int


class GeographicImpactQuery(TypedDict):
    schema_version: Literal["butterflylens-geographic-impact-query:v1.0.0"]
    project_id: str
    accepted_taxon_keys: list[str]
    snapshot_mode: Literal["live", "submitted"]
    h3_resolution: int
    scope: dict[str, str]
    event_date_from: str | None
    event_date_to: str | None
    evidence_maturity: list[str]
    ala_basis_of_record: list[str]
    review_states: list[str]
    page_size: int
    query_fingerprint: str

File: packages/contracts/src/geographic-impact.ts
export const GEOGRAPHIC_IMPACT_CELL_SCHEMA_VERSION =
  'butterflylens-geographic-impact-cell:v1.0.0' as const
export const GEOGRAPHIC_IMPACT_SNAPSHOT_SCHEMA_VERSION =
  'butterflylens-geographic-impact-snapshot:v1.0.0' as const
export const GEOGRAPHIC_IMPACT_QUERY_SCHEMA_VERSION =
  'butterflylens-geographic-impact-query:v1.0.0' as const

export const EVIDENCE_COUNT_STATUSES = [
  'available',
  'unavailable',
  'withheld',
  'not_applicable',
] as const

export type EvidenceCountStatus = (typeof EVIDENCE_COUNT_STATUSES)[number]
export type SnapshotMode = 'live' | 'submitted'

export interface EvidenceCount {
  readonly status: EvidenceCountStatus
  readonly value: number | null
  readonly reason: string | null
}

export interface ImpactFlag {
  readonly status: 'available' | 'unavailable'
  readonly value: boolean | null
  readonly reason: string | null
}

export interface GeographicImpactCell {
  readonly schema_version: typeof GEOGRAPHIC_IMPACT_CELL_SCHEMA_VERSION
  readonly cell_id: string
  readonly grid: 'H3'
  readonly h3_version: string
  readonly h3_resolution: number
  readonly project_id: string
  readonly run_id: string
  readonly snapshot_mode: SnapshotMode
  readonly accepted_taxon_key: string
  readonly ala_snapshot_fingerprint: string | null
  readonly flickr_snapshot_fingerprint: string | null
  readonly provider_union_fingerprint: string | null
  readonly review_projection_fingerprint: string | null
  readonly quality_snapshot_fingerprint: string | null
  readonly counts: {
    readonly ala_baseline: EvidenceCount
    readonly flickr_candidate: EvidenceCount
    readonly yoloe_butterfly: EvidenceCount
    readonly bioclip_species_candidate: EvidenceCount
    readonly community_reviewed: EvidenceCount
    readonly human_supported: EvidenceCount
    readonly release_ready: EvidenceCount
  }
  readonly impact: {
    readonly potential_coverage_gap: ImpactFlag
    readonly human_supported_additional: ImpactFlag
    readonly release_ready_additional: ImpactFlag
  }
  readonly nearest_ala_evidence_distance: {
    readonly status: 'available' | 'unavailable' | 'not_applicable'
    readonly metres: number | null
    readonly reason: string | null
  }
  readonly latest_ala_event_date: string | null
  readonly latest_flickr_event_date: string | null
  readonly data_deficiency_state:
    | 'baseline_present'
    | 'candidate_only'
    | 'baseline_and_candidate'
    | 'no_eligible_evidence'
    | 'baseline_unavailable'
    | 'coordinates_withheld'
    | 'insufficient_precision'
  readonly public_geometry: {
    readonly status: 'available' | 'generalized' | 'withheld'
    readonly source_precision_metres: number | null
    readonly published_h3_resolution: number | null
    readonly reason: string | null
  }
  readonly evidence_fingerprints: readonly string[]
  readonly cell_fingerprint: string
  readonly scientific_claim_allowed: false
}

export interface GeographicImpactSnapshot {
  readonly schema_version: typeof GEOGRAPHIC_IMPACT_SNAPSHOT_SCHEMA_VERSION
  readonly snapshot_id: string
  readonly project_id: string
  readonly run_id: string
  readonly mode: SnapshotMode
  readonly country_code: 'AU'
  readonly status: 'available' | 'stale' | 'unavailable'
  readonly generated_at: string
  readonly last_updated_at: string
  readonly submitted_source_commit: string | null
  readonly worker_heartbeat_fingerprint: string | null
  readonly cell_schema_version: typeof GEOGRAPHIC_IMPACT_CELL_SCHEMA_VERSION
  readonly cell_count: number
  readonly cell_artifact_checksum: string
  readonly cell_artifact_fingerprint: string
  readonly query_fingerprint: string
  readonly map_projection_fingerprint: string
  readonly blockers: readonly string[]
  readonly append_only_revision: number
}

export interface GeographicImpactQuery {
  readonly schema_version: typeof GEOGRAPHIC_IMPACT_QUERY_SCHEMA_VERSION
  readonly project_id: string
  readonly accepted_taxon_keys: readonly string[]
  readonly snapshot_mode: SnapshotMode
  readonly h3_resolution: number
  readonly scope: {
    readonly kind: 'national' | 'state' | 'territory' | 'ibra' | 'lga' | 'h3'
    readonly scope_id: string
  }
  readonly event_date_from: string | null
  readonly event_date_to: string | null
  readonly evidence_maturity: readonly string[]
  readonly ala_basis_of_record: readonly string[]
  readonly review_states: readonly string[]
  readonly page_size: number
  readonly query_fingerprint: string
}
